### PR TITLE
Handle toDataURL with no context

### PR DESCRIPTION
**Pull request by dzbarsky**
[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8725)

### CODE REVIEW

File: components/script/dom/htmlcanvaselement.rs [hunk 1]
```diff
@@ -279,7 +279,7 @@ impl HTMLCanvasElementMethods for HTMLCanvasElement {
             let encoded = encoded.to_base64(STANDARD);
             Ok(DOMString::from(format!("data:{};base64,{}", mime_type, encoded)))
         } else {
-            Err(Error::NotSupported)
+            Ok(DOMString::from(format!("data:image/png;")))
```
Comment (eefriedman) [inline components/script/dom/htmlcanvaselement.rs]: "When its canvas context mode is none, a canvas element has no rendering context, and its bitmap must be fully transparent black with an intrinsic width equal to the numeric value of the element's width attribute and an intrinsic height equal to the numeric value of the element's height attribute, those values being interpreted in CSS pixels, and being updated as the attributes are set, changed, or removed." (https://html.spec.whatwg.org/multipage/scripting.html#the-canvas-element).

In other words, a canvas always has content, so we must always return a valid image with the specified width and height.

**Comment by dzbarsky** [comment]
Good catch. We're matching Chrome's output now.

**Comment by eefriedman** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[components/script/dom/htmlcanvaselement.rs, line 271 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8725#-K4GLUtdM8c4DAmIxnia:-K4GLUtdM8c4DAmIxnib:-1310202380) ([raw file](https://github.com/servo/servo/blob/73956ac93208d76a54d2f2b12a9d1ea5259e2343/components/script/dom/htmlcanvaselement.rs#L271)):*
Please make this explicitly `None`, and add an explicit branch for any unhandled case.  I think handling for webgl contexts is missing?

---

*[components/script/dom/htmlcanvaselement.rs, line 272 \[r2\]](https://reviewable.io:443/reviews/servo/servo/8725#-K4GLqQ-eIisL9ptPboH:-K4GLqQ-eIisL9ptPboI:-1453817404) ([raw file](https://github.com/servo/servo/blob/73956ac93208d76a54d2f2b12a9d1ea5259e2343/components/script/dom/htmlcanvaselement.rs#L272)):*
If I'm not mistaken, this is solid white, and the spec says "fully transparent black".  At the very least, there should be a comment explicitly noting what color this is.

Also, a test for this behavior would be nice.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8725)*

**Comment by dzbarsky** [comment]
Don't have time to write a test right now, but if you mark me delegate+ and I'll land with a test.

**Comment by eefriedman** [comment]
I'd like to review the test to double-check that it make sense.

Also, this doesn't build as-is; looks like a bug in the lint.  I'll look into it.

**Comment by dzbarsky** [comment]
Fixed the test @eefriedman

File: components/script/dom/htmlcanvaselement.rs [hunk 2]
```diff
@@ -258,39 +258,42 @@ impl HTMLCanvasElementMethods for HTMLCanvasElement {
...
-
-            let mut encoded = Vec::new();
-            {
-                let encoder: PNGEncoder<&mut Vec<u8>> = PNGEncoder::new(&mut encoded);
-                encoder.encode(&raw_data, self.Width(), self.Height(), ColorType::RGBA(8)).unwrap();
-            }
+        // Step 2.
+        if self.Width() == 0 || self.Height() == 0 {
+            return Ok(DOMString::from("data:,"));
+        }
```
Comment (eefriedman) [inline components/script/dom/htmlcanvaselement.rs]: Does step 2 need to come before step 3?

Otherwise looks fine.

@bors-servo delegate+

Comment (dzbarsky) [inline components/script/dom/htmlcanvaselement.rs]: Not sure if GetImageData can actually throw here, but I rearranged the steps just to be safe.

**Pull request merged**